src/main/java/org/testng/DependencyMap.java

```diff
@@ -82,13 +83,22 @@ private static boolean hasInstance(
       ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
     Object baseInstance = baseClassMethod.getInstance();
     Object derivedInstance = derivedClassMethod.getInstance();
-    return derivedInstance != null || baseInstance != null;
+    boolean result = derivedInstance != null || baseInstance != null;
+    boolean params = null != baseClassMethod.getFactoryMethodParamsInfo() && null != derivedClassMethod.getFactoryMethodParamsInfo().getParameters();
+    if (result && params && System.getProperty("testng.thread.affinity", "false").equals("true")) {
+      return baseClassMethod.getFactoryMethodParamsInfo().getParameters()[0] == derivedClassMethod.getFactoryMethodParamsInfo().getParameters()[0];
```

Comment:
Shouldn't we be doing an equality check rather than checking a reference equality ?